transport/x/cherami/inbound.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cherami
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/sync"
+	"go.uber.org/yarpc/serialize"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/cherami-client-go/client/cherami"
+)
+
+const (
+	transportName = "cherami"
+
+	defaultPrefetchCount = 10
+
+	defaultCheramiTimeoutInSec = 15
+)
+
+// InboundConfig defines the config in order to create a Inbound
+// if frontend is provided, we'll connect to the provided frontend
+// otherwise, hyperbahn will be used to connect to cherami
+// PrefetchCount controls the number of messages to buffer locally.
+// Inbounds which process messages very fast may want to specify larger value
+// for PrefetchCount for faster throughput.  On the flip side larger values for
+// PrefetchCount will result in more messages being buffered locally causing high memory foot print
+
+type InboundConfig struct {
+	Destination         string
+	ConsumerGroup       string
+	Frontend            string
+	Port                int
+	PrefetchCount       int
+	CheramiTimeoutInSec int
+}
+
+// Inbound is a inbound that uses cherami as the transport
+type Inbound struct {
+	config         InboundConfig
+	consumer       cherami.Consumer
+	router         transport.Router
+	tracer         opentracing.Tracer
+	cheramiFactory CheramiFactory
+
+	once sync.LifecycleOnce
+}
+
+// NewInbound builds a new cherami inbound
+func NewInbound(config InboundConfig) *Inbound {
```

Comment:
so the idea behind a "Transport" object is to consolidate all the logic that is transport specific.
In TChannel, this means storing the tchannel.Channel object.
In HTTP this means storing the http.Client.

This has a few benefits by allowing us to have one place to create objects to share across inbounds/outbounds.

So with the current design if we want to add the same object across inbounds/outbounds it will look something like this:

```go
logger := NewLogger()
metrics := NewMetrics()
tracer := NewTracer()
inbounds := []Inbounds{
  cherami.NewInbound(cherami.InboundConfig{
    logger: logger,
    metrics: metrics,
    tracer: tracer,
    ...
  },
  cherami.NewInbound(cherami.InboundConfig{
    logger: logger,
    metrics: metrics,
    tracer: tracer,
    ...
  },
  ...
}
outbounds := []Outbounds{
  cherami.NewOutbound{
    logger: logger,
    metrics: metrics,
    tracer: tracer,,
    ...
  },
  ...
}
```

By centralizing the Transport-Specific concerns into a "Transport" object we can simplify this API by setting "global" objects on the transport, and using the transport to create the inbounds and outbounds:


```go
cheramiTransport := cherami.NewTransport()
cheramiTransport.SetLogger(NewLogger())
cheramiTransport.SetMetrics(NewMetrics())
cheramiTransport.SetTracer(NewTracer())

inbounds := []Inbounds{
  cheramiTransport.NewInbound(cherami.InboundConfig{...}},
  cheramiTransport.NewInbound(cherami.InboundConfig{...}},
  ...
}
outbounds := []Outbounds{
  cheramiTransport.NewOutbound{cherami.OutboundConfig{...}},
  ...
}
```